### ISSUE TITLE
Gauge of Boxen usage and problems

### ISSUE THREAD

**Issue by jacobbednarz** [boxen/boxen#213]
Hello!

Over the last 6 or so months I've been working through various incentives for Boxen to bring the maintenance and outstanding issues under wraps. I think this has been quite successful so far and I'm starting to get a clearer picture of where the pain points are and what needs more ❤️ . 

Even though this picture is becoming more filled in for me, I'm interested to hear from you around the the pain points you've hit with Boxen and what sort of usage it is still getting at your workplaces (or even if it's just you!). I'm also quite interested from those that have moved away from Boxen, their reasoning and what sort of solutions you are using now.

Thanks!

cc @rafaelfranca @MikeMcQuaid @indirect @fromonesrc @patrickod @n0ts @grosser @jhaals @lucasmazza @seanknox @salimane @petems @typhonius

**Comment by grosser** [comment]
We moved away from it ... just a few stragglers just using it until they get a new laptop ...
Mostly using a docker based solution new with some custom plumbing to get everything working well together and a custom cli to start/restart/connect things.

**Comment by MikeMcQuaid** [comment]
I [wrote a blog post](http://mikemcquaid.com/2016/06/15/replacing-boxen/) about moving GitHub internally from Boxen to our new setup with [Strap](https://github.com/MikeMcQuaid/strap/), [Homebrew Bundle](https://github.com/Homebrew/homebrew-bundle) and [GitHub’s “scripts to rule them all” pattern](https://github.com/github/scripts-to-rule-them-all). GitHub no longer uses Boxen internally.

I've not been paying close attention to Boxen so it could be that some or all of these have been improved or fixed but here's my laundry list from maintaining, using and supporting Boxen at GitHub:
- there was a significant amount of overhead to update a package to a new version. It involved copying a file from Homebrew, modifying it, generating new binary packages, uploading then, creating a new tag, updating the release in our local Boxen, testing that, pushing it. That's maybe reasonable for major version updates of critical tools but it became a massive hassle for security updates that need rolled out quickly. This is why in Homebrew's new versions support we're not allowing people to sit on patch versions; I'm generally critical of the "pin all your dependencies" approach when it means people often sit on insecure versions until they need a newer one. In GitHub we have bots that handle this for us but it's an industry-wide annoyance.
- related to the prior: there's a lot of duplication of effort and work in Boxen compared to Homebrew. Things get fixed faster in Homebrew than in Boxen because more people use it. Boxen should lean on Homebrew and other Homebrew ecosystem tools whenever possible rather than reimplementing the wheel.
- Puppet is not well suited to systems it does not have complete control over. If you change things under Puppet it seems to go 💥 relatively easily.
- Writing Puppet is less pleasant than writing Ruby
- Creating and maintaining CI at GitHub for Boxen was a pain. Builds were extremely time consuming. Incremental builds were not a reliable indicator of first-install behaviour.
- Related to the above: new installs bitrot much quicker than existing installs.
- Even technical end-users don't understand Puppet's behaviour or error-messaging. As a result there was some degree of learned helplessness when it came to either making Boxen do what you wanted it to do or when things went wrong.
- A no-op Boxen run took an extremely long time.
- It makes more sense to have project configuration live in project repositories rather than a single repository.
- I don't personally like that Boxen makes it easy for Boxen administrators to decide that all sorts of crap should be installed on my machine. If you employ e.g. mobile and web developers chances are you end up installing unnecessary stuff for one or both groups because that's how Boxen generally nudges you towards. Instead, dependencies should be projects specific with a very small subset of systemwide dependencies (e.g. Homebrew on a Mac).

In short, I think both the implementation and administrator of Boxen are overly complex compared to what you can build on top of existing tooling.

---

I appreciate few or none of the above may apply widely to other Boxen users and that some of them may not be fixable but figured you'd rather I was candid about this. Thanks for you work on Boxen and in open source @jacobbednarz: the project is lucky to have you 💖

**Comment by petems** [comment]
I think @MikeMcQuaid summarises it pretty well.

I've worked with Puppet for about 6 years, worked at Puppet Inc for over 2 years now, I've maintained Boxen at 2 and a half organisations. Managing that Boxen installation was probably like 40% of my day sometimes (large government project, lots of new starters, most were not big Ruby or Puppet people).

The main problem is you're cooking several fairly tricky things together in one big stew: Puppet, Homebrew and people's personal configurations.

Some of it was a product of it's time, homebrew cask didn't support binary installs, and it could be a lot flakier back in the day (2011-2012), error messages weren't always super easy to parse.

Added to that, you have standard bitrot, random URLs from the internet for packages that 404'd and disappeared and things were fixed slower because of the smaller install-base of Boxen over homebrew

Since then, Homebrew is basically a swiss army knife that does pretty much whatever you need to setup a development Apple machine, it's a lot more stable, and seems to be ticking along super smoothly, mainly due to the hardwork of @MikeMcQuaid and the Homebrew team (I still have my Homebrew kickstarter sticker proudly on my home machine 😄)

I actually still use Boxen myself, mainly because I've got it _just right_ but maintaining something like that for a big org was a massive headache. 

I think it was a great way of getting the developer system configuration movement started, but for a lot of people it's a bit overkill for what people need. Strap is a great alternative, and hits the key points of what most people need.

I think a place for Boxen still exists, especially when organisations have more complex requirements.  
But for most people, Strap + Brew bundle files + Dotfiles is what you really need. 

I disagree with this however...

> Writing Puppet is less pleasant than writing Ruby

but hey, opinions 😉

**Comment by jacobbednarz** [comment]
Thanks for the feedback @grosser @petems @MikeMcQuaid! I really appreciate the time taken to convey it all. 

> There's a lot of duplication of effort and work in Boxen compared to Homebrew. Things get fixed faster in Homebrew than in Boxen because more people use it. Boxen should lean on Homebrew and other Homebrew ecosystem tools whenever possible rather than reimplementing the wheel.

This was one of the primary drivers for #203. I noticed that a great deal of modules and pain came from vendoring our own versions of formulae and decided it was no longer beneficial. From my chats with the original Boxen maintainers, this was originally done because Homebrew was less reliable then now and didn't ship with features like enforced checksums for all formulae which could have resulted in security issues. Since then, @MikeMcQuaid and Co have done a tremendous job at getting everything working smoothly and quite reliable which made it far easier to just rely on instead of duplicating it. I'm still looking to cut a bunch of other bits and bobs but for now, it's far lighter on the Boxen side than ever before.

> Since then, Homebrew is basically a swiss army knife that does pretty much whatever you need to setup a development Apple machine, it's a lot more stable, and seems to be ticking along super smoothly, mainly due to the hardwork of @MikeMcQuaid and the Homebrew team (I still have my Homebrew kickstarter sticker proudly on my home machine 😄)

Major +1 on this. 👏  @MikeMcQuaid and Homebrew maintainers.

> Creating and maintaining CI at GitHub for Boxen was a pain. Builds were extremely time consuming. Incremental builds were not a reliable indicator of first-install behaviour.

:crossed_fingers: that boxen/our-boxen#835 is the start of better CI for Boxen. I understand testing OSX has been a pain in the past and I'm looking to clearly outline some better paths for testing. I'd like to get to the point where unit testing is done for each push and then before merging, a full integration test run is kicked off with just that module installed for the `our-boxen`. This will give us the ability to know A) whether a module is doing what it says on the tin and B) plays nicely when installed with the rest of Boxen.

> A no-op Boxen run took an extremely long time.

This speed issue is caused by a few things:

- The amount Boxen need to install. Unfortunately many of the "fixing" modules (like openssl) can be slow to compile and install. We've been periodically reviewing this list and hoping to cull some more in the future.
- Boxen is still on Puppet 3. There are some performance wins to be had by upgrading to Puppet 4 but I haven't got around to it just yet (#209).

> It makes more sense to have project configuration live in project repositories rather than a single repository.

This is a tricky one for me because things like our local MySQL configuration is managed by Boxen yet the `database.yml` lives in the repository and needs values from the Puppet manifest. I'm still unsure how to tackle this one at the moment.

> Instead, dependencies should be projects specific with a very small subset of systemwide dependencies (e.g. Homebrew on a Mac).

I'm not sure whether this is an issue with Boxen or the specific implementation you've used. The way we've structured projects at Envato works nicely because we've built out a bunch of "projects" that you add to your personal manifest as you move between teams and when you don't need it anymore, you just remove it and it cleans up whatever it installed. 

> Added to that, you have standard bitrot

I'm hoping to add a feature in the future that will force a Boxen run every _n_ days. Most of the rot issues I've seen have been due to people not maintaining their local setup and eventually it just gets too far gone to maintain sanely. It seems people mostly think their local environments are a set and forget type of issue (which IMO is never the case) and once you install everything once, you never maintain it again. This isn't so much a Boxen issue but would be nice to tackle in an automated fashion.

Thanks again for the feedback and keep it coming!

**Comment by MikeMcQuaid** [comment]
> From my chats with the original Boxen maintainers, this was originally done because Homebrew was less reliable then now and didn't ship with features like enforced checksums for all formulae which could have resulted in security issues. Since then, @MikeMcQuaid and Co have done a tremendous job at getting everything working smoothly and quite reliable which made it far easier to just rely on instead of duplicating it. I'm still looking to cut a bunch of other bits and bobs but for now, it's far lighter on the Boxen side than ever before.

A general comment on open source here is that both at the time and now: Homebrew (and, really, any project) can't fix problems it doesn't now about. If you (or indeed anyone in this thread) have problems like this with Homebrew: please report issues so we can solve them or at least to start a conversation towards finding a solution. There's a bunch of issues Boxen worked around in Homebrew that we would have been happy to fix (or, ideally, could have been fixed in Homebrew rather than Boxen).

Thanks for the kind words, too!

> I'd like to get to the point where unit testing is done for each push and then before merging, a full integration test run is kicked off with just that module installed for the our-boxen. This will give us the ability to know A) whether a module is doing what it says on the tin and B) plays nicely when installed with the rest of Boxen.

Sounds like a great idea. In my experience with stuff like Boxen/Homebrew: integration tests are a pain but that's where most of the real issues are found.

> This is a tricky one for me because things like our local MySQL configuration is managed by Boxen yet the database.yml lives in the repository and needs values from the Puppet manifest. I'm still unsure how to tackle this one at the moment.

For Boxen specifically I'd say it'd be good if there more of the configuration could live in the repository itself and Boxen can set that up if it does the cloning or a script can be called which calls back to Boxen if users do the cloning (I'd suggest supporting both paths).

> I'm not sure whether this is an issue with Boxen or the specific implementation you've used.

Yes, quite possibly a specific implementation.

> It seems people mostly think their local environments are a set and forget type of issue (which IMO is never the case) and once you install everything once, you never maintain it again. This isn't so much a Boxen issue but would be nice to tackle in an automated fashion.

I think local environments should be set and forget unless you're specifically doing something e.g. like starting a server for a given project. Ideally then scripts would call back into Boxen to ask if anything needs to be done and have a very fast no-op when they do not (i.e. <1s). I think it's a problem to rely on users to manually run `boxen` periodically but equally to run it automatically when they may find the automation results in a broken install. People are loath to change what's already working for them.
